ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,113 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.mozilla.telemetry.ingestion.core.Constant.Attribute;
+import com.mozilla.telemetry.transforms.PubsubConstraints;
+import com.mozilla.telemetry.util.Time;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessageWithAttributesCoder;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.transforms.WithTimestamps;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Instant;
+
+/**
+ * Count the number of impressions aggregated by reporting URL.
+ */
+public class AggregateImpressions
+    extends PTransform<PCollection<PubsubMessage>, PCollection<PubsubMessage>> {
+
+  private static final List<String> aggregationFields = ImmutableList.of(
+      ParsedReportingUrl.PARAM_COUNTRY_CODE, ParsedReportingUrl.PARAM_REGION_CODE,
+      ParsedReportingUrl.PARAM_FORM_FACTOR, ParsedReportingUrl.PARAM_OS_FAMILY,
+      ParsedReportingUrl.PARAM_ID);
+
+  private final ValueProvider<String> aggregationWindowDuration;
+
+  public AggregateImpressions(ValueProvider<String> aggregationWindowDuration) {
+    this.aggregationWindowDuration = aggregationWindowDuration;
+  }
+
+  public static AggregateImpressions of(ValueProvider<String> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages //
+        // Add reporting url as key, pubsub message as value
+        .apply(WithKeys.of((SerializableFunction<PubsubMessage, String>) //
+        AggregateImpressions::getAggregationKey)) //
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), PubsubMessageWithAttributesCoder.of())) //
+        // Set timestamp to current time
+        .apply(WithTimestamps.of(message -> new Instant())) //
+        // Group impressions into timed windows
+        // TODO: Use valueprovider
+        .apply("IntervalWindow", Window.into(FixedWindows.of(Time.parseDuration("10m")))) //
```

Comment:
For this line if I do `Time.parseDuration(aggregationWindowDuration.get())`, I get this error 
```
Value only available at runtime, but accessed from a non-runtime context: RuntimeValueProvider{propertyName=providerRuntimeValues, default=null}
```
 in the unit test that has `pipeline.apply(AggregateImpressions.of(pipeline.newProvider("10m")))` but not in the real pipeline.  I understand the error but what would be a good workaround here?

Reply:
It looks like FixedWindows doesn't give us any hook to use a ValueProvider, so it expects to know window configuration at compile time. I don't see any way around that, which means `aggregationWindowDuration` needs to become a compile-time parameter.

This whole problem goes away once we're able to move to deploys based on Flex Templates.